### ISSUE TITLE
Inconsistent f-string output for rounded np.float32

### ISSUE THREAD

**Issue by davidmascharka** [numpy/numpy#10645]
NumPy 1.13.3, Python 3.6.2, Ubuntu 16.04

```python
>>> str(round(1337.09997559, 1))
'1337.1'

>>> f'{round(1337.09997559, 1)}'
'1337.1'

>>> f'{round(np.float32(1337.09997559), 1)}'
'1337.0999755859375'

>>> str(round(np.float32(1337.09997559), 1))
'1337.1'

>>> f'{round(np.float64(1337.09997559), 1)}'
'1337.1'

>>> str(round(np.float64(1337.09997559), 1))
'1337.1'
```

**Comment by seberg** [comment]
This should be nicer in 1.14 which uses Dragon4 (I think) to make the floating point precision string only as long as necessary to be exact. Note that the output is not wrong, just more convoluted then expected.

**Comment by charris** [comment]
Looks like the float32 value of round is getting converted to float64 before string conversion. With current master:
```
>>> a = round(np.float32(1337.09997559), 1)
>>> a
1337.1
>>> '%22.15f' % a
'  1337.099975585937500'
>>> b = round(np.float64(1337.09997559), 1)
>>> '%22.15f' % b
'  1337.099999999999909'
```

**Comment by seberg** [comment]
Yeah, nvm. Maybe not fixed, will reopen for now (though I am not sure if it is an issue or not right away)

**Comment by charris** [comment]
It may be a python issue. The `repr` for the `float32` is `1337.1`.

**Comment by eric-wieser** [comment]
Under the hood the f string falls back on `__format__` - which I think converts to a python float and passes on the format spec to python.

**Comment by eric-wieser** [comment]
Yep:

https://github.com/numpy/numpy/blob/maintenance/1.14.x/numpy/core/src/multiarray/scalartypes.c.src#L358-L426

In particular, the following lines:

https://github.com/numpy/numpy/blob/maintenance/1.14.x/numpy/core/src/multiarray/scalartypes.c.src#L398-L400

Which cast the `np.float32` to `float`.

**Comment by eric-wieser** [comment]
An easy fix would be to implement `__format__` as `str` for the case when there is no format string - like we did in #9883.

**Comment by eric-wieser** [comment]
Simpler repro on master:
```python
>>> f = np.float32(1337.0999755859375)
>>> format(f)
'1337.0999755859375'
>>> str(f)
'1337.1'
```
